### DISCUSSION TITLE
Syslog Replacement

### DISCUSSION THREAD

**Discussion by tastyfrankfurt** [vectordotdev/vector#8155] [General]
Hey Guys,

We mainly deal with centos/ubuntu/redhat/rocky.  These come with a default installation of rsyslog.  Currently trying to work out a suitable configuration to replace the defaults outputs of these.  So far i have struggled to get my head around the vrl implementation.  Is someone able to help?  Filtering the log is not an issue its the outputing in a format that is usable to the log files, ie data host processname proc id then the message.  We are using journald as the source.

Thanks in advanced, i have also attached the outputs i am looking at.
*.info;mail.none;authpriv.none;cron.none                /var/log/messages
authpriv.*                                              /var/log/secure
mail.*                                                  -/var/log/maillog
cron.*                                                  /var/log/cron
*.emerg                                                 :omusrmsg:*
uucp,news.crit                                          /var/log/spooler
local7.*                                                /var/log/boot.log

**Comment by spencergilbert** [comment]
If you're just looking to parse the messages that are syslog formatted you can do something like:
```rust
parsed, err = parse_syslog(.message)
if err != null {
  log("Message not parsed: " + string!(.message), level: "error")
}
. = merge(., parsed)
```

**Reply by tastyfrankfurt** [reply]
Sorry i think maybe i need to re ask the question, i want to uninstall rsyslog and use vector as a replacement to write to the previous log file /var/log/messages,/var/log/secure etc.  At the moment i can get vector to source from journald and filter on the log types to get the messages in the right place, the messages however, are missing timestamps, host, process, processid, message.  The logs coming off journald are in json format.
So one of my attempts was below, where i filter, then try to remap, i guess you could do this in any order.
  t_messages_filter:
    type: filter
    inputs: [s_journald]
    condition: includes(["6"], .PRIORITY) && !includes(["2","4","10","9"], .SYSLOG_FACILITY)
  t_messages_remap:
    type: remap
    inputs: [s_journald]
    source: join!([.timestamp," ",get_hostname!()," ",.message])

**Reply by spencergilbert** [reply]
Could you share a sample event you're pulling off the `journald` source?

**Reply by tastyfrankfurt** [reply]
Sure can and thanks for taking the time to look at this, been doing my head in, certainly as i get better at the vrl this should become easier for me, currently working on routing for an aggregator so matching/filtering/adding/dropping fields is straight forward is just the message munging, same with most syslog implementations, the munging is always the ticky part.  See sample events below from Journald.

{"PRIORITY":"6","SYSLOG_FACILITY":"3","SYSLOG_IDENTIFIER":"vector","_BOOT_ID":"015e44c61d8f41ad94d38415fa84e3e0","_CAP_EFFECTIVE":"400","_CMDLINE":"/usr/bin/vector --config /etc/vector/vector.yaml --config /etc/vector/conf.d/*.yaml","_COMM":"vector","_EXE":"/usr/bin/vector","_GID":"988","_MACHINE_ID":"0787888c3632445d999ec22b7f7aba1e","_PID":"1200405","_SELINUX_CONTEXT":"system_u:system_r:unconfined_service_t:s0","_STREAM_ID":"9f42bac8c00743af905be22b9163585a","_SYSTEMD_CGROUP":"/system.slice/vector.service","_SYSTEMD_INVOCATION_ID":"80bc21ad2f7f4af6b7f203360f8d5a74","_SYSTEMD_SLICE":"system.slice","_SYSTEMD_UNIT":"vector.service","_TRANSPORT":"stdout","_UID":"1001","__MONOTONIC_TIMESTAMP":"7720428206571","__REALTIME_TIMESTAMP":"1625657062303667","host":"hostA","message":"Jul 07 11:24:22.303  INFO vector::topology: Removing sink. name=\"d_file\"","source_type":"journald","timestamp":"2021-07-07T11:24:22.303667Z"}
{"PRIORITY":"6","SYSLOG_FACILITY":"3","SYSLOG_IDENTIFIER":"vector","_BOOT_ID":"015e44c61d8f41ad94d38415fa84e3e0","_CAP_EFFECTIVE":"400","_CMDLINE":"/usr/bin/vector --config /etc/vector/vector.yaml --config /etc/vector/conf.d/*.yaml","_COMM":"vector","_EXE":"/usr/bin/vector","_GID":"988","_MACHINE_ID":"0787888c3632445d999ec22b7f7aba1e","_PID":"1200405","_SELINUX_CONTEXT":"system_u:system_r:unconfined_service_t:s0","_STREAM_ID":"9f42bac8c00743af905be22b9163585a","_SYSTEMD_CGROUP":"/system.slice/vector.service","_SYSTEMD_INVOCATION_ID":"80bc21ad2f7f4af6b7f203360f8d5a74","_SYSTEMD_SLICE":"system.slice","_SYSTEMD_UNIT":"vector.service","_TRANSPORT":"stdout","_UID":"1001","__MONOTONIC_TIMESTAMP":"7720428206571","__REALTIME_TIMESTAMP":"1625657062303667","host":"hostA","message":"Jul 07 11:24:22.303  INFO vector::topology: Running healthchecks.","source_type":"journald","timestamp":"2021-07-07T11:24:22.303667Z"}
{"PRIORITY":"6","SYSLOG_FACILITY":"3","SYSLOG_IDENTIFIER":"vector","_BOOT_ID":"015e44c61d8f41ad94d38415fa84e3e0","_CAP_EFFECTIVE":"400","_CMDLINE":"/usr/bin/vector --config /etc/vector/vector.yaml --config /etc/vector/conf.d/*.yaml","_COMM":"vector","_EXE":"/usr/bin/vector","_GID":"988","_MACHINE_ID":"0787888c3632445d999ec22b7f7aba1e","_PID":"1200405","_SELINUX_CONTEXT":"system_u:system_r:unconfined_service_t:s0","_STREAM_ID":"9f42bac8c00743af905be22b9163585a","_SYSTEMD_CGROUP":"/system.slice/vector.service","_SYSTEMD_INVOCATION_ID":"80bc21ad2f7f4af6b7f203360f8d5a74","_SYSTEMD_SLICE":"system.slice","_SYSTEMD_UNIT":"vector.service","_TRANSPORT":"stdout","_UID":"1001","__MONOTONIC_TIMESTAMP":"7720428206571","__REALTIME_TIMESTAMP":"1625657062303667","host":"hostA","message":"Jul 07 11:24:22.303  INFO vector::topology: Starting sink. name=\"d_file02\"","source_type":"journald","timestamp":"2021-07-07T11:24:22.303667Z"}
{"PRIORITY":"6","SYSLOG_FACILITY":"3","SYSLOG_IDENTIFIER":"vector","_BOOT_ID":"015e44c61d8f41ad94d38415fa84e3e0","_CAP_EFFECTIVE":"400","_CMDLINE":"/usr/bin/vector --config /etc/vector/vector.yaml --config /etc/vector/conf.d/*.yaml","_COMM":"vector","_EXE":"/usr/bin/vector","_GID":"988","_MACHINE_ID":"0787888c3632445d999ec22b7f7aba1e","_PID":"1200405","_SELINUX_CONTEXT":"system_u:system_r:unconfined_service_t:s0","_STREAM_ID":"9f42bac8c00743af905be22b9163585a","_SYSTEMD_CGROUP":"/system.slice/vector.service","_SYSTEMD_INVOCATION_ID":"80bc21ad2f7f4af6b7f203360f8d5a74","_SYSTEMD_SLICE":"system.slice","_SYSTEMD_UNIT":"vector.service","_TRANSPORT":"stdout","_UID":"1001","__MONOTONIC_TIMESTAMP":"7720428206571","__REALTIME_TIMESTAMP":"1625657062303667","host":"hostA","message":"Jul 07 11:24:22.303  INFO vector::topology: Starting sink. name=\"d_file01\"","source_type":"journald","timestamp":"2021-07-07T11:24:22.303667Z"}
{"PRIORITY":"6","SYSLOG_FACILITY":"3","SYSLOG_IDENTIFIER":"vector","_BOOT_ID":"015e44c61d8f41ad94d38415fa84e3e0","_CAP_EFFECTIVE":"400","_CMDLINE":"/usr/bin/vector --config /etc/vector/vector.yaml --config /etc/vector/conf.d/*.yaml","_COMM":"vector","_EXE":"/usr/bin/vector","_GID":"988","_MACHINE_ID":"0787888c3632445d999ec22b7f7aba1e","_PID":"1200405","_SELINUX_CONTEXT":"system_u:system_r:unconfined_service_t:s0","_STREAM_ID":"9f42bac8c00743af905be22b9163585a","_SYSTEMD_CGROUP":"/system.slice/vector.service","_SYSTEMD_INVOCATION_ID":"80bc21ad2f7f4af6b7f203360f8d5a74","_SYSTEMD_SLICE":"system.slice","_SYSTEMD_UNIT":"vector.service","_TRANSPORT":"stdout","_UID":"1001","__MONOTONIC_TIMESTAMP":"7720428206571","__REALTIME_TIMESTAMP":"1625657062303667","host":"hostA","message":"Jul 07 11:24:22.303  INFO vector::topology::builder: Healthcheck: Passed.","source_type":"journald","timestamp":"2021-07-07T11:24:22.303667Z"}
{"PRIORITY":"6","SYSLOG_FACILITY":"3","SYSLOG_IDENTIFIER":"vector","_BOOT_ID":"015e44c61d8f41ad94d38415fa84e3e0","_CAP_EFFECTIVE":"400","_CMDLINE":"/usr/bin/vector --config /etc/vector/vector.yaml --config /etc/vector/conf.d/*.yaml","_COMM":"vector","_EXE":"/usr/bin/vector","_GID":"988","_MACHINE_ID":"0787888c3632445d999ec22b7f7aba1e","_PID":"1200405","_SELINUX_CONTEXT":"system_u:system_r:unconfined_service_t:s0","_STREAM_ID":"9f42bac8c00743af905be22b9163585a","_SYSTEMD_CGROUP":"/system.slice/vector.service","_SYSTEMD_INVOCATION_ID":"80bc21ad2f7f4af6b7f203360f8d5a74","_SYSTEMD_SLICE":"system.slice","_SYSTEMD_UNIT":"vector.service","_TRANSPORT":"stdout","_UID":"1001","__MONOTONIC_TIMESTAMP":"7720428207025","__REALTIME_TIMESTAMP":"1625657062304121","host":"hostA","message":"Jul 07 11:24:22.303  INFO vector: Vector has reloaded. path=[(\"/etc/vector/conf.d/client.yaml\", None), (\"/etc/vector/vector.yaml\", None)]","source_type":"journald","timestamp":"2021-07-07T11:24:22.304121Z"}

**Reply by spencergilbert** [reply]
And to be clear you want to convert the json formatted messages to syslog, like
```
<13>1 2020-03-13T20:45:38.119Z dynamicwireless.name non 2426 ID931 [exampleSDID@32473 iut="3" eventSource= "Application" eventID="1011"] Try to override the THX port, maybe it will reboot the neural interface!
```

**Reply by tastyfrankfurt** [reply]
So using the json and taking the existing key value pair above something like below, (note the timestamp is converted to something close to BSD)
<timestamp> <hostfield> <_COMM>[<_PID>]: <message>
Jul 10 07:08:35 hostA.example.local vector[1200405]: Jul 07 11:24:22.303 INFO vector::topology: Removing sink. name="d_file"
Jul 10 07:08:35 hostA.example.local vector[1200405]: Jul 07 11:24:22.303 INFO vector::topology: Running healthchecks."
Jul 10 07:08:35 hostA.example.local vector[1200405]: Jul 07 11:24:22.303 INFO vector::topology: Starting sink. name="d_file02"
Jul 10 07:08:35 hostA.example.local vector[1200405]: Jul 07 11:24:22.303 INFO vector::topology: Starting sink. name="d_file01"

**Reply by spencergilbert** [reply]
A quick script it threw together that you could use before a `file` sink with `encoding = text`:

```rust
message = string!(.timestamp) + " " + string!(.host) + " " + string!(.SYSLOG_IDENTIFIER) + "[" + string!(._PID) + "]: " + string!(.message)
```

I didn't convert the timestamp but that does appear to come out how you'd like

**Reply by tastyfrankfurt** [reply]
Oh man was so close on previous attempts will give this a go and let you know how i went.  Thanks @spencergilbert

**Reply by tastyfrankfurt** [reply]
Hey @spencergilbert  got it done, will write up a config but this is what i went with
.message = format_timestamp!(now(),format: "%b %e %H:%M:%S") + " " + string!(.host) + " " + string!(.SYSLOG_IDENTIFIER) + "[" + string!(._PID) + "]: " + string!(.message)

**Reply by tastyfrankfurt** [reply]
Okay have written up a configuration that we will use for ubuntu and centos, its important that the vector user is a member of the syslog group on ubuntu, haven't tested redhat/centos yet.  But this work, thanks for getting me the info i needed @spencergilbert, working copy for us to replace rsyslog completely with vector. 

---
sources:
  s_journald:
    type: journald
    batch_size: {{ vector_client_journald_batchsize }}
    current_boot_only: true
    exclude_units: {{ vector_client_journald_units_exclude }}
    journalctl_path: /usr/bin/journalctl
transforms:
  t_remap_message:
    type: remap
    inputs: [s_journald]
    source: |
      if (!is_null(.host) && !is_null(.SYSLOG_IDENTIFIER) && !is_null(._PID) && !is_null(.message))
      {
        .message = format_timestamp!(now(),format: "%b %e %H:%M:%S") + " " + string!(.host) + " " + string!(.SYSLOG_IDENTIFIER) + "[" + string!(._PID) + "]: " + string!(.message)
        .drop = "fields not null"
      }
  t_local_filter:
    type: route
    inputs: [t_remap_message]
    route:
      "t_messages": includes(["6"], .PRIORITY) && !includes(["2","4","10","9"], .SYSLOG_FACILITY) && exists(.drop)
      "t_secure": includes(["10"], .SYSLOG_FACILITY) && exists(.drop)
      "t_mail": includes(["2"], .SYSLOG_FACILITY) && exists(.drop)
      "t_cron": includes(["9"], .SYSLOG_FACILITY) && exists(.drop)
      "t_emerg": includes(["8"], .SYSLOG_FACILITY) || (includes(["58"], .PRIORITY) && !includes(["7"], .SYSLOG_FACILITY)) && exists(.drop)
      "t_local": includes(["23"], .SYSLOG_FACILITY) && exists(.drop)
sinks:
  d_messages:
    type: file
    inputs: [ t_local_filter.t_messages ]
    compression: none
    path: "/var/log/messages"
    encoding:
      codec: "text"
  d_secure:
    type: file
    inputs: [ t_local_filter.t_secure ]
    compression: none
    path: "/var/log/secure"
    encoding:
      codec: "text"
  d_mail:
    type: file
    inputs: [ t_local_filter.t_mail ]
    compression: none
    path: "/var/log/maillog"
    encoding:
      codec: "text"
  d_cron:
    type: file
    inputs: [ t_local_filter.t_cron ]
    compression: none
    path: "/var/log/cron"
    encoding:
      codec: "text"
  d_emerg:
    type: file
    inputs: [ t_local_filter.t_emerg ]
    compression: none
    path: "/var/log/spooler"
    encoding:
      codec: "text"
  d_local:
    type: file
    inputs: [ t_local_filter.t_local ]
    compression: none
    path: "/var/log/boot.log"
    encoding:
      codec: "text"

**Reply by spencergilbert** [reply]
Neat! If you wanted to make the remap more robust you could replace the `!` ([raising the error](https://vector-dot-dev.netlify.app/docs/reference/vrl/errors/#raising)), with a `result, err` pattern and better inform you if one of those fields isn't in the expected format - though coming off of journald I expect you should be pretty safe.

**Reply by spencergilbert** [reply]
Related - https://github.com/timberio/vector/issues/7117 would make it easier/cleaner to set up that message string you want

**Reply by tastyfrankfurt** [reply]
Yes indeed, but happy for now, my next challenge is to work out log rotate, dont suppose a simply nohup will support this otherwise could be looking at a full service restart eeeeeek
